// rangen/dist/user.js
"use strict";
/* tslint:disable:no-magic-numbers */
Object.defineProperty(exports, "__esModule", { value: true });
var random_1 = require("./random");
var number_1 = require("./number");
var boolean_1 = require("./boolean");
var id_1 = require("./id");
var female_first_names_1 = require("./helpers/female-first-names");
var male_first_names_1 = require("./helpers/male-first-names");
var last_names_1 = require("./helpers/last-names");
exports.user = function (params) {
    var count = (params && params.count) ? params.count : 1;
    var gen = (params && params.gender) ? params.gender : 'any';
    var minA = (params && params.minAge) ? params.minAge : 18;
    var maxA = (params && params.maxAge) ? params.maxAge : 55;
    var results = [];
    for (var i = 0; i < count; i++) {
        var gender = gen === 'any' ? random_1.random(['male', 'female']) : gen;
        var first = random_1.random(gender === 'male' ? male_first_names_1.maleFirstNames : female_first_names_1.femaleFirstNames);
        var last = random_1.random(last_names_1.lastNames);
        var age = number_1.num(minA, maxA);
        results.push({
            gender: gender,
            name: {
                title: gender === 'male' ? 'mr' : 'mrs',
                first: first,
                last: last
            },
            email: (first + "." + last + "@example.com").toLowerCase(),
            age: age,
            dob: getDob(age),
            registered: getRegistered(),
            phone: number_1.num(100, 999) + "-" + number_1.num(100, 999) + "-" + number_1.num(1000),
            id: id_1.id()
            // location:
            // username: '',
            // password: '',
            // nat: '',
            // picture: {}
        });
    }
    return results;
};
function getDob(age) {
    function daysInMonth(y, m) {
        return new Date(y, m, 0).getDate();
    }
    var date = new Date();
    // already had BD this year
    var year = date.getFullYear() - age;
    var month = number_1.num(0, date.getMonth());
    var day = number_1.num(1, daysInMonth(year, month + 1));
    if (month === date.getMonth()) {
        day = number_1.num(1, date.getDate());
    }
    // did not have BD this year
    if (boolean_1.bool()) {
        year = year - 1;
        month = number_1.num(date.getMonth(), 11);
        day = number_1.num(1, daysInMonth(year, month + 1));
        if (month === date.getMonth()) {
            day = number_1.num(date.getDate(), daysInMonth(year, month + 1));
        }
    }
    return new Date(year, month, day);
}
function getRegistered() {
    // from 1min to 3years
    var timestamp = number_1.num(1000 * 60, 60 * 60 * 24 * 365 * 1000 * 3);
    var now = new Date().getTime();
    return new Date(now - timestamp);
}
